src/common/Gateway.sol

```diff
@@ -96,7 +100,13 @@ contract Gateway is Auth, Recoverable, IGateway {
             bytes memory message = remaining.slice(0, length);
             remaining = remaining.slice(length, remaining.length - length);
 
-            try processor_.handle(centrifugeId, message) {
+            uint256 remainingGas = gasleft() - GAS_FAIL_MESSAGE_STORAGE;
+            uint256 executionGas = gasService.messageGasLimit(localCentrifugeId, message);
+
+            // We only want to fail inside `processor.handle()` by logic stuff, not by out of gas issues
+            require(remainingGas >= executionGas, NotEnoughGasToProcess());
+
+            try processor_.handle{gas: remainingGas}(centrifugeId, message) {
```

Comment:
Important note, we only need to double check there is enough gas to process the message, but we must leave all the gas available to processor.handle() passing remainingGas.

Not used gas is useful again after `processor.handle()`